File: common/src/main/java/com/pluxity/permission/PermissionGroup.java
package com.pluxity.permission;

import com.pluxity.global.entity.BaseEntity;
import com.pluxity.user.entity.RolePermission;
import jakarta.persistence.*;
import java.util.HashSet;
import java.util.Set;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "permission_group")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class PermissionGroup extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    private String description;

    @OneToMany(mappedBy = "permissionGroup", cascade = CascadeType.ALL)
    private final Set<RolePermission> rolePermissions = new HashSet<>();

    @OneToMany(mappedBy = "permissionGroup", cascade = CascadeType.ALL)
    private final Set<Permission> permissions = new HashSet<>();

    @Builder
    public PermissionGroup(String name, String description) {
        this.name = name;
        this.description = description;
    }

    public void changeName(String name) {
        this.name = name;
    }

    public void changeDescription(String description) {
        this.description = description;
    }

    public void addPermission(Permission permission) {
        if (permission != null && !this.permissions.contains(permission)) {
            this.permissions.add(permission);
            if (permission.getPermissionGroup() != this) {
                permission.changePermissionGroup(this);
            }
        }
    }

    public void removePermission(Permission permission) {
        if (permission != null && this.permissions.contains(permission)) {
            this.permissions.remove(permission);
            if (permission.getPermissionGroup() == this) {
                permission.clearPermissionGroup();
            }
        }
    }

    public void clearPermissions() {
        new HashSet<>(this.permissions).forEach(this::removePermission);
    }
}

File: common/src/main/java/com/pluxity/permission/PermissionGroup.kt
package com.pluxity.permission

import com.pluxity.global.entity.BaseEntity
import com.pluxity.user.entity.RolePermission
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import jakarta.persistence.Table

@Entity
@Table(name = "permission_group")
class PermissionGroup(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,
    var name: String,
    var description: String?,
) : BaseEntity() {
    @OneToMany(mappedBy = "permissionGroup", cascade = [CascadeType.ALL])
    val rolePermissions: MutableSet<RolePermission> = HashSet()

    @OneToMany(mappedBy = "permissionGroup", cascade = [CascadeType.ALL])
    val permissions: MutableSet<Permission> = HashSet()

    fun changeName(name: String) {
        this.name = name
    }

    fun changeDescription(description: String) {
        this.description = description
    }

    fun addPermission(permission: Permission) {
        if (!permissions.contains(permission)) {
            permissions.add(permission)
            if (permission.permissionGroup != this) {
                permission.changePermissionGroup(this)
            }
        }
    }

    fun removePermission(permission: Permission) {
        if (permissions.contains(permission)) {
            permissions.remove(permission)
            if (permission.permissionGroup == this) {
                permission.clearPermissionGroup()
            }
        }
    }
}
